clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -0,0 +1,1021 @@
+        return (int) deltaMs;
+    }
+
+    /**
+     * Generate the client id based on the configuration.
+     *
+     * @param config                            The configuration
+     *
+     * @return                                  The client id
+     */
+    static String generateClientId(AdminClientConfig config) {
+        String clientId = config.getString(AdminClientConfig.CLIENT_ID_CONFIG);
+        if (clientId.length() > 0)
+            return clientId;
+        return "adminclient-" + ADMIN_CLIENT_ID_SEQUENCE.getAndIncrement();
+    }
+
+    /**
+     * Get the deadline for a particular call.
+     *
+     * @param optionTimeoutMs                   The timeout option given by the user.
+     *
+     * @return                                  The deadline in milliseconds.
+     */
+    private long getDeadlineMs(Integer optionTimeoutMs) {
+        if (optionTimeoutMs != null)
+            return time.milliseconds() + Math.min(0, optionTimeoutMs);
+        // If optionTimeoutMs is less than 0, we use the default timeout.
+        return time.milliseconds() + defaultTimeoutMs;
+    }
+
+    /**
+     * Pretty-print an exception.
+     *
+     * @param throwable                         The exception.
+     *
+     * @return                                  A compact human-readable string.
+     */
+    static String prettyPrintException(Throwable throwable) {
+        if (throwable == null)
+            return "Null exception.";
+        if (throwable.getMessage() != null) {
+            return throwable.getClass().getSimpleName() + ": " + throwable.getMessage();
+        }
+        return throwable.getClass().getSimpleName();
+    }
+
+    /**
+     * Close an AutoCloseable, suppressing any exceptions.
+     *
+     * @param what                              Human-readable string explaining what we are closing.
+     * @param closeable                         What we are closing, or null to ignore.
+     */
+    static void closeQuietly(String what, AutoCloseable closeable) {
+        if (closeable == null)
+            return;
+        try {
+            closeable.close();
+        } catch (Throwable exc) {
+            log.error("Error closing " + what, exc);
+        }
+    }
+
+    static KafkaAdminClient create(AdminClientConfig config) {
+        Metadata metadata = null;
+        Metrics metrics = null;
+        NetworkClient networkClient = null;
+        Time time = Time.SYSTEM;
+        String clientId = generateClientId(config);
+        ChannelBuilder channelBuilder = null;
+        Selector selector = null;
+        ApiVersions apiVersions = new ApiVersions();
+
+        try {
+            metadata = new Metadata(config.getLong(AdminClientConfig.RETRY_BACKOFF_MS_CONFIG),
+                    config.getLong(AdminClientConfig.METADATA_MAX_AGE_CONFIG));
+            metrics = new Metrics(new MetricConfig(), new LinkedList<MetricsReporter>(), time);
+            String metricGrpPrefix = "adminclient";
```

Comment:
I don't feel strongly about it either way.  It gets used in Selector.java to build the metric group name as follows: `String metricGrpName = metricGrpPrefix + "-metrics";`

I'm not aware of any convention that says we can't have a hyphen here, so I guess I'll put it in...